rosetta-source/src/main/rosetta/product-qualification-func.rosetta

```diff
@@ -261,34 +277,38 @@ func Qualify_CreditDefaultSwaption: <"This product qualification is temporary un
         economicTerms -> payout -> optionPayout only exists
             and Qualify_AssetClass_Credit(
-                    economicTerms -> payout -> optionPayout only-element -> underlier -> contractualProduct -> economicTerms
+                    economicTerms -> payout -> optionPayout only-element -> optionUnderlier -> Product ->> economicTerms
                 ) = True
 
-func Qualify_UnderlierProduct_Equity: <"Qualifies a product as having the Asset Class classification Equity.">
+func Qualify_UnderlierObservable_Equity: <"Qualifies an Observable as having the Asset Class classification Equity.">
     inputs:
-        underlier Product (1..1)
+        underlier Observable (1..1)
     output:
         is_product boolean (1..1)
 
     set is_product:
-        underlier -> security -> securityType = SecurityTypeEnum -> Equity
-            or (underlier -> security -> securityType = SecurityTypeEnum -> Fund
-                and underlier -> security -> fundType = FundProductTypeEnum -> ExchangeTradedFund)
-            or (underlier -> security -> securityType = SecurityTypeEnum -> Fund
-                and underlier -> security -> fundType = FundProductTypeEnum -> MutualFund)
-            or underlier -> security -> securityType = SecurityTypeEnum -> Warrant
-            or underlier -> index ->> assetClass = AssetClassEnum -> Equity
+        Qualify_SecurityTypeEquity(underlier -> Asset -> Instrument -> Security)
+            or underlier -> Index ->> assetClass = AssetClassEnum -> Equity
             // Qualifies that the underlier is a basket composed of equity products only
-            or (underlier -> basket exists
-                and (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Equity
-                    or (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Fund
-                        and underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> fundType any = FundProductTypeEnum -> ExchangeTradedFund)
-                    or (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Fund
-                        and underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> fundType any = FundProductTypeEnum -> MutualFund)
-                    or underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Warrant
-                    or underlier -> basket -> basketConstituent -> index ->> assetClass any = AssetClassEnum -> Equity 
+            or (underlier -> Basket exists
+                and (Qualify_SecurityTypeEquity(underlier -> Basket -> basketConstituent -> Asset -> Instrument -> Security)
+                    or underlier -> Basket -> basketConstituent -> Index ->> assetClass all = AssetClassEnum -> Equity 
```

Comment:
In the `Basket` case, replace these `or` statements with a single line calling `Qualify_UnderlierObservable_Equity` on each `basketConstituent` (as single cardinality argument) and check that all are True.

Reply:
I can't get this to work as `Qualify_UnderlierObservable_Equity` needs an `Observable` and this object is not accessible in `basketConstituent` (though may be it should be?)

Reply:
You _may_ be able to pass `basketConstituent` directly to a function that expects an `Observable` (limited implementation of inheritance).

Reply:
it works (at least syntactically)